arbnode/espresso_batcher_addr_monitor.go

```diff
@@ -0,0 +1,154 @@
+package arbnode
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+const (
+	eventKey = "espresso-batcher-addr"
+)
+
+// `BatchPosterSet` event
+type BatcherAddrEvent struct {
+	l1Height  uint64         `koanf:"l1-height"`
+	addr      common.Address `koanf:"addr"`
+	isBatcher bool           `koanf:"is-batcher"`
+}
+
+type BatcherAddrMonitorInterface interface {
+	AddBatchPosterSetEvent(l1 uint64, addr common.Address, isBatcher bool) error
+	// Returns the valid batcher addresses up to the given L1 height
+	GetValidAddresses(l1 uint64) []common.Address
+	// The latest L1 height that has been seen by the monitor. This implies
+	// that all events up to this height have been processed.
+	SetL1Height(l1 uint64)
+}
+
+var _ BatcherAddrMonitorInterface = (*BatcherAddrMonitor)(nil)
+
+type BatcherAddrMonitor struct {
+	l1Height          uint64
+	lastEventL1Height uint64
+	cached            bool
+	// Cache for the latest valid addresses.
+	// Since batcher address changes are infrequent and callers typically
+	// process HotShot blocks sequentially, caching improves performance.
+	cachedAddresses []common.Address
+
+	events        []BatcherAddrEvent
+	db            ethdb.Database
+	initAddresses []common.Address
+}
+
+func NewBatcherAddrMonitor(
+	addr []common.Address,
+	db ethdb.Database,
+) *BatcherAddrMonitor {
+	return &BatcherAddrMonitor{
+		initAddresses: addr,
+		db:            db,
+	}
+}
+
+func (b *BatcherAddrMonitor) AddBatchPosterSetEvent(l1 uint64, addr common.Address, isBatcher bool) error {
+	event := BatcherAddrEvent{
+		l1Height:  l1,
+		addr:      addr,
+		isBatcher: isBatcher,
+	}
+	b.events = append(b.events, event)
+	// Sort events by l1Height to ensure correct processing order.
+	// Since BatcherAddr events are infrequent, the performance impact of sorting is negligible.
+	sort.Slice(b.events, func(i, j int) bool {
+		return b.events[i].l1Height < b.events[j].l1Height
+	})
+	b.lastEventL1Height = l1
+	b.cached = false
+	return b.Store()
+}
+
+func (b *BatcherAddrMonitor) GetValidAddresses(l1 uint64) []common.Address {
```

Comment:
where is this function being called in the code?

Reply:
In the espresso streamer, where it attempts to parse hotshot payload